### ISSUE TITLE
NumSigFigs Odd Behavior

### ISSUE THREAD

**Issue by dhilsen** [maths/moodle-qtype_stack#256]
I have encountered a strange issue with a particular NumSigFigs [a,0] test in one of our questions and wondered if anyone might have some insight into what is happening. We are utilizing the latest ATUnitsNum release, and when testing this particular PRT containing NumSigFigs [a,0] it does not seem to recognize trailing zeros after the decimal as being significant. It would almost seem that an older version of NumSigFigs is being utilized. Oddly enough, the very same question uses NumSigFigs [a,0] in a different PRT and it works perfectly fine. (This particular question was authored at an earlier date prior to the latest NumSigFigs upgrade, and did originally contain the NumSigFigs answer test which at the time had problems with trailing zeros.)

Has anyone experienced something similar to this in the past?

**Comment by sangwinc** [comment]
Are you using the cache for CAS results?

I can't reproduce this bug with the current unit tests.  Is the faulty case included here?
https://github.com/maths/moodle-qtype_stack/blob/master/tests/fixtures/answertestfixtures.class.php

**Comment by dhilsen** [comment]
The NumSigFigs [a,0] case that is failing does appear in the Answer Tests script you referenced. This is extremely odd as this same NumSigFigs test appears in a different PRT within the same question and is working perfectly fine.

It's not clear to me how it might be determined if we are using the cache for CAS results. Could you describe how this might be checked?

**Comment by sangwinc** [comment]
You define whether you are using the cache in the stack settings page

Site Administration -> plungins -> question type -> stack.

You can clear the cache from the healthcheck page (if you are using the cache).

Chris

**Comment by aharjula** [comment]
If you are using a POST cache like nginx in front of a MaximaPool or cluster of them you will need to clear that cache using whatever tools that cache has...

**Comment by dhilsen** [comment]
Thank you for the input on clearing the cache. The cache was cleared on the STACK healthcheck page and "Purge all caches" from Moodle site administration was also performed. The NumSigFigs [a,0] test still failed to recognize trailing zeros after the decimal as being significant. 

The problem was resolved by deleting the entire PRT and then recreating it. After doing this the NumSigFigs [a,0] test recognized trailing zeros as being significant and worked properly. It would appear that somehow the original PRT was utilizing an older version of the NumSigFigs test. (As mentioned earlier, this particular question was authored at an earlier date prior to the latest NumSigFigs upgrade, and did originally contain the NumSigFigs answer test which at the time had problems with trailing zeros.)

I'm curious if anyone might have some idea as to what caused this problem. The concern is that other previously authored questions using the NumSigFigs test might also be failing, and in order to get a better handle on this we would need to test each of the several hundred questions using this test.

**Comment by sangwinc** [comment]
Gosh!  I'm honestly not sure how that situation would have arisen.  Unless there is some caching in Moodle I'm not aware about.  Tim, can you suggest anything here?

**Comment by timhunt** [comment]
As far as I am aware, the only caching in STACK is the CAS caching that you clear on the healthcheck page. (And, that is not affected by Moodle's purge all caches button.)

When I have this sort of issue, I find trying to reproduce it on the CAS chat page is a good thing to do.

**Comment by dhilsen** [comment]
It certainly is possible this problem is unique to the one particular question. However, it might be constructive to understand exactly what happened here. Chris, would it be helpful it I emailed the question to you? (I still have a copy of the original failing version.)

**Comment by sangwinc** [comment]
Yes, please do email me the question and I'll look.  I want to eradicate any behaviour like this....

**Comment by sangwinc** [comment]
Thank you for tracking this down.  This was an annoying bug caused by not "triming" whitespace from variable names.

In fact, I got so annoyed by having to put in "print_r" statements to track this down I've exposed a lot more of the trace through the PRT in the question testing page.  This will help question authors see the expressions which are going into each answer test and will make debugging questions a lot easier.
